### ISSUE TITLE
Provide an official ARM docker image and/or provide documentation for building on ARM

### ISSUE THREAD

**Issue by gizmochief7** [envoyproxy/envoy#1861]
Would it be possible to get an official Docker image of Envoy on ARM or at least some documentation on how to build it?
My team and I have tried off and on to build Envoy on ARM using the Bazel build process without any luck. Its challenging to say the least. I just saw a PR #1795 by @costinm dated a couple weeks ago based on the issue #1781 that talked about building it on ARM successfully after a few fixes. This is great news but we aren't sure how to duplicate it. @costinm mentioned that he used a **custom CMAKE and a pi-cross-compiler** to make it work. Documentation on this would be very helpful. Also, if there exist an image I could use for testing, then that would be very helpful too. Thanks again for such an awesome product!

**Comment by vielmetti** [comment]
@gizmochief7 it's been a little while, have you made progress? It sounds like you're building for 32-bit ARMv7; my particular interest is in arm64 builds.

**Comment by gizmochief7** [comment]
@vielmetti It has been and I almost forgot about posting this. We did not have any further luck with this attempt and without any community help we decided to move past Envoy. We tried out another reverse proxy/side car project for micro services and it just worked out of the box on both 32-bit and 64-bit. So we haven't wasted any more time here. Sorry I couldn't be more help.

**Comment by rshriram** [comment]
fwiw, I tried to build bazel on my pi (yes, it took several hours), and then built envoy on pi for armv7l 64bit. While envoy started up and was able to talk to http services, it had difficulty talking to certain https services (esp twitter). There seemed to be some boringssl issue that prevented ssl connections from working. My hunch was that it didn't compile properly. At that point, I gave up, because I had spent 4 solid days fighting bazel on pi.

**Comment by costinm** [comment]
I think the cross-compilation option is faster and less painful - and easy
to expand to other platforms.

Boringssl and most of the dependencies have native support for cmake (or
can be easily adapted) - one of the problems with building with
'bazel' on pi is that deps are not really built with bazel, so even if
bazel can cross-compile c++ code, it would still not help us, most of the
code comes from the libraries compiled with autoconf. Of course autoconf
also supports cross - but then I also wanted to run it on android,
and cmake is supported very nicely by the NDK.

It requires some periodic tweaking - adding the new envoy files and fixing
ocasional 32-bit regressions (so far only the flags package seems to be a
problem - not clear why), but it's a pretty fast build.

But I appreciate the pain and effort in straight-compiling and fighting
bazel :-)

**Comment by costinm** [comment]
https://github.com/costinm/istio-build/releases  if you want to try some
(older) binaries, matching ~ Istio 0.2

**Comment by moderation** [comment]
I've secured a Pine64 Rock64 SOC which is ARM8 /  aarch64 and has 4GB of RAM. I've successfully 
compiled Bazel 0.13.0. I'm not having any luck compiling Envoy.

I've leveraged @clnperez [way](https://github.com/clnperez/envoy-ppc64le/blob/master/WORKSPACE#L19-L21) of removing the Lua JIT.

I can't get her trick to use the locally installed Go to work `go_register_toolchains(go_version="host")`. When I use this Bazel uses a symbolic link to the installed Go tools (`~/Library/go/pkg/tool/linux_arm64`) and tries to compile Gazelle and protoc-gen-go and fails because `link` can't see `gcc` on $PATH. Installing these tools on the host don't fix the issue. I expect this is something to do with breaking the hermetic seal by using `go_version="host"`.

I've modified `source/extensions/extensions_build_config.bzl` to only include a minimum number of extensions.

I'm using `bazel --bazelrc=/dev/null build -s -c opt //source/exe:envoy-static.stripped --define google_grpc=disabled --define signal_trace=disabled --define hot_restart=disabled --verbose_failures` as the build command with the following WORKSPACE.

```
workspace(name = "envoy")

load("//bazel:repositories.bzl", "envoy_dependencies")
load("//bazel:cc_configure.bzl", "cc_configure")

envoy_dependencies(
   skip_targets=['luajit']
)
cc_configure()

load("@envoy_api//bazel:repositories.bzl", "api_dependencies")
api_dependencies()

load("@io_bazel_rules_go//go:def.bzl", "go_download_sdk", "go_rules_dependencies", "go_register_toolchains")
load("@com_lyft_protoc_gen_validate//bazel:go_proto_library.bzl", "go_proto_repositories")

go_proto_repositories(shared=0)
go_rules_dependencies()

go_download_sdk(name = "go_sdk", sdks = {"linux_arm64":  ("go1.10.2.linux-arm64.tar.gz", "d6af66c71b12d63c754d5bf49c3007dc1c9821eb1a945118bfd5a539a327c4c8"), }, )

go_register_toolchains()
```
This prevents the build automatically downloading Go for `linux_amd64`.

The build then fails with 

```
ERROR: ~/.cache/bazel/_bazel_moderation/1d277edd85b03195d275adc4870a9bef/external/com_lyft_protoc_gen_validate/validate/BUILD:35:1: no such package '@com_github_golang_protobuf//ptypes/duration': no such package '@io_bazel_rules_go_repository_tools//': no such package '@go_sdk//': Unsupported host linux_amd64 and referenced by '@com_lyft_protoc_gen_validate//validate:go_default_library'
ERROR: Analysis of target '//source/exe:envoy-static.stripped' failed; build aborted: no such package '@com_github_golang_protobuf//ptypes/duration': no such package '@io_bazel_rules_go_repository_tools//': no such package '@go_sdk//': Unsupported host linux_amd64
```
My suspicion is that the [lyft/protoc-gen-validate](https://github.com/lyft/protoc-gen-validate/) is hard coded for linux_amd64 (or considering compilation works on MacOS / Darwin maybe it just doesn't support ARM).

Any help from Bazel wizards like @jmillikin-stripe or from people that have tried non amd64 compilation like @clnperez @costinm @rshriram would be great.

**Comment by clnperez** [comment]
I'll try and help. I don't have much ARM experience but a coworker does. So we'll try to put our heads together on this one. Any progress we can make would be great. We'd also like to have a ppc64le docker image. 

wonder twins unite? 

My first guess is that something in cc_configure works differently for ARM. I ran into some things where, when using certain flags, gcc would behave differently for power than it did for x86. envoy, for the last release, seems to have come to rely a lot more on how bazel configured the gcc toolchain and did a little less customization. That got rid of almost all the issues I was seeing, but maybe there is some more tweaking needed for ARM.

For your WORKSPACE file, what's the logic behind including the call to `go_download_sdk'?

**Comment by moderation** [comment]
Before resuming work on this I think we should prioritize moving to rules_go 0.12.0. At the moment a build with 0.12.0 fails with a `cycle in dependency graph` error. I suspect the root cause for this and the inability to build on aarch64 is lyft/protoc-gen-validate. It has a number of hacks and workarounds in that it works today but I think without some maintenance it will prevent a move to rules_go 0.12.0 and the ability to compile for non amd64, darwin, s390x platforms.

**Comment by clnperez** [comment]
So, I'm hitting this now as well on ppc64le. :D

**Comment by moderation** [comment]
@clnperez I've had success this week building on aarch64 / arm64 using these patches from @lubinsz https://github.com/envoyproxy/envoy/pull/3681#issuecomment-400208276. @htuch unblocked protobuf 3.6.0 with this patch to lyft/protoc-gen-validate https://github.com/lyft/protoc-gen-validate/commit/345b6b478ef955ad31382955d21fb504e95f38c7. 0.12.0 is still blocked.

**Comment by clnperez** [comment]
@moderation i just took a look at those patches, and they look like they're all adding arm support, whereas we already have power support. i'm actually hitting a different error today than I did last night, and that is the gazelle compile (at least i think it's the gazelle compile) isn't finding gcc in my PATH. I might open a new issue for that.

**Comment by moderation** [comment]
@clnperez We were having the same issue on arm64. The patch for rules_go at https://github.com/lubinsz/rules_go/commit/497b48893fb04f195dfa8d81ec4201265e91d0d0 fixed that issue. Not sure how portable that patch is but might give you some ideas.

**Comment by clnperez** [comment]
@moderation using it as-is seems to break compatibility with lyft/protoc-gen-validate as the patch for rules_go is based off of a branch that deprecated `go_repositories`. so i tried just changing the path bits for the linking (using what was in that patch) and it didn't get me around that. 

there are so many moving parts to trying out one thing, though, i could have missed something. but i *think* i have it all lined up correctly. i'll dig some more and see what i find.

**Comment by clnperez** [comment]
@moderation it turns out this was what i needed: https://github.com/bazelbuild/bazel-gazelle/pull/242

**Comment by moderation** [comment]
@clnperez this is similar to this patch that enabled the arm64 build - https://github.com/lubinsz/rules_go/commit/497b48893fb04f195dfa8d81ec4201265e91d0d0#diff-b88d1275b3ee8565f7821ba59b402139. Thanks @lubinsz. It would be good to go back and have these changes included upstream.

**Comment by vielmetti** [comment]
Several upstream issues in `bazel` have been resolved, specifically https://github.com/bazelbuild/rules_go/pull/1550 which might make this worth a retry.

**Comment by bryanburgers** [comment]
With the work in bazelbuild/rules_go#1550 #4809 #5285 #5308, it looks like we're getting closer. What still needs to be done?

I'm trying to build an embedded device that exposes gRPC to external clients, and a simple controlling website via gRPC-Web for when an external client isn't available, so having Envoy for Arm would be awesome.

**Comment by lafonj** [comment]
I am very interested in this as well. If there is any way I can help, I would love to help push this through.

**Comment by den-vasyliev** [comment]
> I am very interested in this as well. If there is any way I can help, I would love to help push this through.

Actually i have spoken devs from Lift on Kubecon 19: envoy on arm in the roadmap and i hope this year officially. But not for today :(

**Comment by moderation** [comment]
@bryanburgers, @lafonj, @den-vasyliev I've been building Envoy on arm64 for quite a while. All of the Bazel and rules_go issues I documented at https://gitlab.com/snippets/1742074have been resolved. I still compile out Lua but no issues otherwise. I compile each new update of Bazel from source when it comes out - no modifications required.

**Comment by den-vasyliev** [comment]
> @bryanburgers, @lafonj, @den-vasyliev I've been building Envoy on arm64 for quite a while. All of the Bazel and rules_go issues I documented at https://gitlab.com/snippets/1742074 have been resolved. I still compile out Lua but no issues otherwise. I compile each new update of Bazel from source when it comes out - no modifications required.

Thanks!

"The minimal extensions_build_config.bzl file that I use to only compile in the features I'm using is at https://gitlab.com/snippets/1742863"
Page Not Found https://gitlab.com/snippets/1742863

**Comment by moderation** [comment]
@den-vasyliev I fixed the snippet (it was private) and I updated it to my current latest. As you can see I compile out a lot of the extensions but it helps with compile time and binary size and there isn't much point including stuff you don't use.

I'd be happy to upload a current binary somewhere. Traveling back from vacation so will be a couple of days until I can update to master and rebuild.

**Comment by den-vasyliev** [comment]
> @den-vasyliev I fixed the snippet (it was private) and I updated it to my current latest. As you can see I compile out a lot of the extensions but it helps with compile time and binary size and there isn't much point including stuff you don't use.
> 
> I'd be happy to upload a current binary somewhere. Traveling back from vacation so will be a couple of days until I can update to master and rebuild.

awesome! welcome back to home!

**Comment by den-vasyliev** [comment]
> I'd be happy to upload a current binary somewhere. Traveling back from vacation so will be a couple of days until I can update to master and rebuild.

just kindly reminder - thx

**Comment by moderation** [comment]
@den-vasyliev I've uploaded a binary to a fork of mine at https://github.com/moderation/envoy/blob/c8a5e574ced826588eeca4649f51ee333967fe7c/envoy_arm64. Keep in mind that I've compiled out a bunch of extensions on this binary. It reports the following at startup.

```
[2019-03-15 15:33:03.395][30807][info][main] [source/server/server.cc:205] statically linked extensions:
[2019-03-15 15:33:03.396][30807][info][main] [source/server/server.cc:207]   access_loggers: envoy.file_access_log
[2019-03-15 15:33:03.396][30807][info][main] [source/server/server.cc:210]   filters.http: envoy.filters.http.rbac,envoy.health_check,envoy.rate_limit,envoy.router
[2019-03-15 15:33:03.396][30807][info][main] [source/server/server.cc:213]   filters.listener: envoy.listener.original_dst,envoy.listener.proxy_protocol,envoy.listener.tls_inspector
[2019-03-15 15:33:03.397][30807][info][main] [source/server/server.cc:216]   filters.network: envoy.filters.network.rbac,envoy.http_connection_manager,envoy.ratelimit,envoy.tcp_proxy
[2019-03-15 15:33:03.397][30807][info][main] [source/server/server.cc:218]   stat_sinks: envoy.statsd
[2019-03-15 15:33:03.397][30807][info][main] [source/server/server.cc:220]   tracers: envoy.lightstep,envoy.zipkin
[2019-03-15 15:33:03.399][30807][info][main] [source/server/server.cc:223]   transport_sockets.downstream: raw_buffer,tls
[2019-03-15 15:33:03.400][30807][info][main] [source/server/server.cc:226]   transport_sockets.upstream: raw_buffer,tls
```

**Comment by AleksandarFilipov** [comment]
@moderation. Thank you for sharing your binary, I'll try and use it. 
This thread is long and there are loads of references which most seem to resolved. However it's still unclear to me where this is going. Could i kindly ask some someone to summarize whats going on.

**Comment by moderation** [comment]
@AleksandarFilipov The instructions on the snippet are now longer required. I will update the snippet to reflect that. Building for aarch64 just works using the latest Bazel and Envoy master. No modifications are needed.

What hasn't been worked out is the automatic building and publishing of aarch64 binaries and container images. The folks at Tetrate may provide these through https://www.tetrate.io/application-tools/getenvoy/ (/cc @zinuga, @pragashj).  There was also some early talk of Packet helping with this also.

**Comment by vielmetti** [comment]
@moderation - there are a couple of CI systems running at Packet that have native support for arm64 builds - Drone, Codefresh, and Shippable. 

I am very glad to hear that the build just works, that's tremendous progress.

**Comment by boubasse** [comment]
Hi there, thank you for that :)

@moderation Do you think you can make a build for armv7?

I'd like to install Istio on a K8s Raspberry Pi Cluster :)

**Comment by moderation** [comment]
@boubasse unfortunately I don't have a way to build for armv7. I don't think compiling Bazel and Envoy will be possible on a Pi given the memory constraints.

**Comment by enmasse** [comment]
Hi all! I created a docker image containing Ubuntu and Envoy for ARM64. I tried to build for ARM32 but I didn't even manage to bootstrap Bazel for ARM32.

https://cloud.docker.com/u/onemasse/repository/docker/onemasse/envoy

**Comment by enmasse** [comment]
Hi all! I created a docker image containing Ubuntu and Envoy for ARM64. I tried to build for ARM32 but I didn't even manage to bootstrap Bazel for ARM32.

https://cloud.docker.com/u/onemasse/repository/docker/onemasse/envoy

**Comment by AleksandarFilipov** [comment]
@enmasse following this guide https://github.com/samjabrahams/tensorflow-on-raspberry-pi/blob/master/GUIDE.md#3-build-bazel (particularly on file changes) and some docker magic seem to do the trick. I'll get back.

**Comment by AleksandarFilipov** [comment]
Failed, but seem fixable, known issue,
https://github.com/bazelbuild/bazel/issues/4483

```
Caused by: java.util.MissingResourceException: Can't find bundle for base name com.google.errorprone.errors, locale en_US
        at java.util.ResourceBundle.throwMissingResourceException(ResourceBundle.java:1573)
        at java.util.ResourceBundle.getBundleImpl(ResourceBundle.java:1396)
        at java.util.ResourceBundle.getBundle(ResourceBundle.java:854)
        at com.sun.tools.javac.util.JavacMessages.lambda$add$0(JavacMessages.java:106)
        at com.sun.tools.javac.util.JavacMessages.getBundles(JavacMessages.java:125)
```

**Comment by vielmetti** [comment]
The CNCF CI team is tackling this at https://github.com/crosscloudci/crosscloudci/issues/134 for cross-reference.

**Comment by Yikun** [comment]
We have able to build the Envoy arm64 binaries in OpenLab, and now, there are two ways to trigger the Envoy arm64 build:

1. [**Periodic  job**](http://status.openlabtesting.org/builds?job_name=envoy-build-arm64): The periodic job had been enabled for a month, now you can see the periodic `envoy-build-arm64` job for `envoyproxy/envoy` in [OpenLab dashboard](http://status.openlabtesting.org/builds?job_name=envoy-build-arm64) [1], the arm build job of `envoyproxy/envoy` master branch will be ran everyday. 
![image](https://user-images.githubusercontent.com/1736354/57911515-92950280-78ba-11e9-937c-0f9c49c5c8bd.png)



2. [**PR triggered job**](https://github.com/theopenlab/envoy/pull/3): you can also do some test on PR triggered in repo [theopenlab/envoy](https://github.com/theopenlab/envoy/pull/3) [2], the demo like [PR](https://github.com/theopenlab/envoy/pull/3) [3],
![image](https://user-images.githubusercontent.com/1736354/57913308-ec97c700-78be-11e9-9e7b-a65e10f24a84.png)


If we want to move the job from theopenlab/envoyinto envoyproxy/envoy officially,  I can help it.



[1] http://status.openlabtesting.org/builds?job_name=envoy-build-arm64
[2] https://github.com/theopenlab/envoy/pull/3

**Comment by Yikun** [comment]
The PR looks like: https://github.com/envoyproxy/envoy/pull/6981

It just likes what we have done in the other CNCF project `containerd` [1] , now each containerd project pr is trigger the build job [2]. 

[1] https://github.com/containerd/containerd/pull/3242
[2] https://github.com/containerd/containerd/pulls

**Comment by aliariff** [comment]
Hey, anyone got luck for running it in raspberry pi?

**Comment by duhruh** [comment]
@aliariff in the same boat, would love to have envoy running on my raspberry pi.

**Comment by lizan** [comment]
With the OpenLab CI configured by @Yikun, you can find their dashboard here http://status.openlabtesting.org/builds?job_name=envoy-build-arm64 and download the binary built there from the artifact. (e.g. https://logs.openlabtesting.org/logs/periodic-4/16/github.com/envoyproxy/envoy/master/envoy-build-arm64/7855faa/test_results/bazel-bin/source/exe).

I'm not sure how well it works, and just FWIW there are number of tests failing in arm64 when I tried in AWS arm64 instance.

**Comment by enmasse** [comment]
@lizan Raspberry Pi runs armhf aka ARM32 aka ARM by default. They're two different architectures. Building ARM64 is manageable. It's not even possible to bootstrap Bazel on ARM32.

**Comment by pranaypratyush** [comment]
I need envoy running on arm7 as well. Tried using qemu by copying `/usr/bin/qemu-x86_64-static` with no success.

**Comment by felix-fabrega** [comment]
@enmasse Why it is not even possible to bootstrap Bazel on arm32 ?
I'm also try harding to cross compile envoy for arm32, on my side right now I'm trying to configure a toolchain for using armv7-eabihf. 
I'm stuck on an error from ninja who seems to not find my compiler binary

**Comment by pranaypratyush** [comment]
I merely need envoy for grpc-web. As an alternate can we have nginx configured on armv7?

**Comment by enmasse** [comment]
Thanks @lizan , I will find the time to try it out.

**Comment by pranaypratyush** [comment]
> I got a PoC cross-compile on `armeabihf` with bazel Raspberry Pi rules ([mjbots/rpi_bazel](https://github.com/mjbots/rpi_bazel)) in the weekend, can anyone of you try it? I don't have a handy device to try right meanwhile.
> 
> Run `bazel build --config=rpi //source/exe:envoy-static` (on x86_64 machine) from this branch: [lizan/envoy@`rpi`](https://github.com/lizan/envoy/tree/rpi) [](/lizan/envoy/tree/HEAD@{2019-08-27T02:55:31Z}) (I have CC=clang / CXX=clang++ in the env FYI)
> 
> And this seems produces a right binary:
> 
> ```
> $ file bazel-bin/source/exe/envoy-static
> bazel-bin/source/exe/envoy-static: ELF 32-bit LSB shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-, for GNU/Linux 3.2.0, BuildID[md5/uuid]=46c4db9ba5e2f0d704073bb757f00943, not stripped
> ```

Could you elaborate on the exact I need to follow? Or better yet give us the binary.

**Comment by dio** [comment]
I tried the following:

```
$ git clone -b rpi git@github.com:lizan/envoy.git
$ cd envoy
$ bazel build --config=rpi --copt -Wno-macro-redefined //source/exe:envoy-static
$ file bazel-bin/source/exe/envoy-static
bazel-bin/source/exe/envoy-static: ELF 32-bit LSB shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-, for GNU/Linux 3.2.0, BuildID[md5/uuid]=1fcaf0ca7143f12c82089c24d935eb08, not stripped
```

It is not stripped so it's about 126MB.

**Comment by lizan** [comment]
Here is a non optimized nor stripped binary, not sure if works.
[envoy-armeabihf.tar.gz](https://github.com/envoyproxy/envoy/files/3557650/envoy-armeabihf.tar.gz)

**Comment by pranaypratyush** [comment]
Was trying to create a docker-image starting from armv7 ubuntu. Any help?

**Comment by pranaypratyush** [comment]
Successfully built the docker-image and ran on odroid-XU4. What tests can I run?

**Comment by pranaypratyush** [comment]
https://hub.docker.com/r/pranaypratyush/envoy-arm32v7

**Comment by enmasse** [comment]
I tried to run it, unsuccessfully. Seems that I have encountered a bug. The link describing how to override the deprecation in the error message also shows 404, so I'm stumped:

```
[2019-09-02 16:35:36.039][6770][info][main] [source/server/server.cc:574] exiting
Proto constraint validation failed (Using deprecated option 'envoy.api.v2.route.CorsPolicy.enabled' from file route.proto. This configuration will be removed from Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/intro/deprecated for details. If continued use of this field is absolutely necessary, see https://www.envoyproxy.io/docs/envoy/latest/configuration/runtime#using-runtime-overrides-for-deprecated-features for how to apply a temporary and highly discouraged override.): allow_origin: "*"
allow_methods: "GET, PUT, DELETE, POST, OPTIONS"
allow_headers: "keep-alive,user-agent,cache-control,content-type,content-transfer-encoding,custom-header-1,x-accept-content-transfer-encoding,x-accept-response-streaming,x-user-agent,x-grpc-web"
expose_headers: "custom-header-1,grpc-status,grpc-message"
max_age: "1728000"
enabled {
  value: true
}

$ ./envoy-armeabihf --version

./envoy-armeabihf  version: 594257a8379d34b0600fd62a5b2f0d9a3f266ec2/1.12.0-dev/Clean/DEBUG/BoringSSL
```

**Comment by lizan** [comment]
@enmasse make sure your config runs on latest on x86_64, i.e. envoyproxy/envoy-dev:6ab225d3c9e24ac22e93d5d12c44b9d4336d0ab4 I think that because cors policy `allow_origin: ` is deprecated. check https://www.envoyproxy.io/docs/envoy/latest/intro/deprecated

**Comment by alexellis** [comment]
/subscribe

@gizmochief7 what was the other tool you used with ARM HF/64 support?

**Comment by dantuluri** [comment]
Did anyone trying using buildx to solve this?

**Comment by pranaypratyush** [comment]
> Did anyone trying using buildx to solve this?

I did. It's been quite some time though. Can't remember the details buildx did not work.

**Comment by pisymbol** [comment]
I'd like to run Ambassador on my pi cloud but it needs Envoy form armv7 first. Has anyone done this before?

**Comment by pranaypratyush** [comment]
> https://hub.docker.com/r/pranaypratyush/envoy-arm32v7

Yes

**Comment by pisymbol** [comment]
@pranaypratyush How did you install?

**Comment by stevef1uk** [comment]
Having got bazel built on a RPi4 and invoking it just so:

bazel build -c opt //source/exe:envoy-static'

I am getting the following errors and am stuck!

INFO: Repository 'com_github_gperftools_gperftools' used the following cache hits instead of downloading the corresponding file.
 * Hash '97f0bc2b389c29305f5d1d8cc4d95e9212c33b55827ae65476fc761d78e3ec5d' for https://github.com/gperftools/gperftools/archive/gperftools-2.7.90.tar.gz
If the definition of 'com_github_gperftools_gperftools' was updated, verify that the hashes were also updated.
**ERROR**: An error occurred during the fetch of repository 'com_github_gperftools_gperftools':
   Traceback (most recent call last):
	File "/home/pi/.cache/bazel/_bazel_pi/528125dac6128c9aedfea67ee71fe51f/external/bazel_tools/tools/build_defs/repo/http.bzl", line 83
		patch(ctx)
	File "/home/pi/.cache/bazel/_bazel_pi/528125dac6128c9aedfea67ee71fe51f/external/bazel_tools/tools/build_defs/repo/utils.bzl", line 165, in patch
		fail(<1 more arguments>)
Error applying patch command ./autogen.sh:
configure.ac:172: error: possibly undefined macro: AC_PROG_LIBTOOL
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
autoreconf: /usr/bin/autoconf failed with exit status: 1
INFO: Call stack for the definition of repository 'boringssl' which is a http_archive (rule definition at /home/pi/.cache/bazel/_bazel_pi/528125dac6128c9aedfea67ee71fe51f/external/bazel_tools/tools/build_defs/repo/http.bzl:292:16):
 - /home/pi/.cache/bazel/_bazel_pi/528125dac6128c9aedfea67ee71fe51f/external/envoy_api/bazel/envoy_http_archive.bzl:17:5
 - /home/pi/.cache/bazel/_bazel_pi/528125dac6128c9aedfea67ee71fe51f/external/envoy_api/bazel/envoy_http_archive.bzl:17:5
 - /home/pi/envoy/bazel/repositories.bzl:16:5
 - /home/pi/envoy/bazel/repositories.bzl:181:5
 - /home/pi/envoy/bazel/repositories.bzl:108:5
 - /home/pi/envoy/WORKSPACE:13:1
ERROR: /home/pi/.cache/bazel/_bazel_pi/528125dac6128c9aedfea67ee71fe51f/external/envoy/bazel/foreign_cc/BUILD:10:1: @envoy//bazel/foreign_cc:gperftools_build depends on @com_github_gperftools_gperftools//:all in repository @com_github_gperftools_gperftools which failed to fetch. no such package '@com_github_gperftools_gperftools//': Traceback (most recent call last):
	File "/home/pi/.cache/bazel/_bazel_pi/528125dac6128c9aedfea67ee71fe51f/external/bazel_tools/tools/build_defs/repo/http.bzl", line 83
		patch(ctx)
	File "/home/pi/.cache/bazel/_bazel_pi/528125dac6128c9aedfea67ee71fe51f/external/bazel_tools/tools/build_defs/repo/utils.bzl", line 165, in patch
		fail(<1 more arguments>)
Error applying patch command ./autogen.sh:
configure.ac:172: error: possibly undefined macro: AC_PROG_LIBTOOL
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
autoreconf: /usr/bin/autoconf failed with exit status: 1
ERROR: Analysis of target '//source/exe:envoy-static' failed; build aborted: Analysis failed
INFO: Elapsed time: 352.106s

**Comment by moderation** [comment]
@stevef1uk Make sure you have autoconf and automake installed. What OS are you using? Bazel version? Are you using the pre-built Clang binaries at https://releases.llvm.org/download.html that allow you to compile with libc++?

I build with `bazel build --jobs 3 --config=libc++ -c opt //source/exe:envoy-static.stripped`

```
cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=19.10
DISTRIB_CODENAME=eoan
DISTRIB_DESCRIPTION="Ubuntu 19.10"
```
```
bazel version
Starting local Bazel server and connecting to it...
Build label: 2.2.0- (@non-git)
Build target: bazel-out/aarch64-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Tue Mar 3 17:24:46 2020 (1583256286)
Build timestamp: 1583256286
Build timestamp as int: 1583256286
```
The `.bazelrc` file I have in my root directory is:
```
cat ~/.bazelrc
build --action_env=CC=clang
build --action_env=CXX=clang++
build --announce_rc
build --config=clang
build --define deprecated_features=disabled
build --define disable_object_dump_on_signal_trace=disabled
build --define hot_restart=disabled
build --define google_grpc=disabled
build --define path_normalization_by_default=true
build --define quiche=enabled
build --define signal_trace=disabled
build --linkopt -fuse-ld=lld
build --override_repository=envoy_build_config=/home/moderation/Library/envoy/build
```
The last line points to my `extensions_build_config.bzl` where I compile out a bunch of extensions I don't need like Thrift, ALTS (inside Google only protocol), AWS, Datadog etc. etc.

**Comment by stevef1uk** [comment]
> @stevef1uk Make sure you have autoconf and automake installed. What OS are you using? Bazel version? Are you using the pre-built Clang binaries at https://releases.llvm.org/download.html that allow you to compile with libc++?

Many thanks. I ensured I had all the required dependencies & the Clang binaries installed and compilation is in progress. 4,322 actions seems to be rather a lot though and it is taking many hours on a RPi4 to compile ;-)

**Comment by lizan** [comment]
We're very close to make ARM64 CI work for building binary. The build image we publish: `envoyproxy/envoy-build-ubuntu:09a5a914c904faa39dbc641181cb43b68cabf626` has arm64 variant so you can use it to build.

Just run `ci/run_envoy_docker.sh 'ci/do_ci.sh bazel.release.server_only'`

**Comment by beriberikix** [comment]
Excellent! Looks like it's using QEMU - does that mean it doesn't need to be built on an Arm machine?

**Comment by lizan** [comment]
We built the ARM build image on QEMU, I guess you can do the full build with QEMU as well but I expect it isn't performant enough so might take long itme.

**Comment by beriberikix** [comment]
Sorry - I didn't mean to use QEMU for the entire build! I interested in building on my local x86 laptop using plain old docker. Am I correct in understanding that since the container is using QEMU to effectively cross-compile, it should build the Arm binary on my x86 machine?

**Comment by beriberikix** [comment]
FWIW, I just tried building `...abf626` on my x86 and the build_release is a x86-64. When I try the same command on a Raspberry Pi 4 running a 64-bit OS I got the error, "docker: no matching manifest for linux/arm/v8 in the manifest list entries."

Note I set the `$IMAGE_ID` to get that specific build. Is there a variable to set to configure aarch64 target while running on an X86?

**Comment by lizan** [comment]
@beriberikix are you running an 32bit OS on your RPi 4? If you're running arm64 you should be able to do it on RPi4

**Comment by beriberikix** [comment]
I'm running the 64 build of Raspbian:
```
pi@pi4:~/envoy $ echo $IMAGE_ID
09a5a914c904faa39dbc641181cb43b68cabf626
pi@pi4:~/envoy $ uname -a
Linux pi4 4.19.97-v8+ #1294 SMP PREEMPT Thu Jan 30 13:27:08 GMT 2020 aarch64 GNU/Linux
pi@pi4:~/envoy $ ci/run_envoy_docker.sh 'ci/do_ci.sh bazel.release.server_only'
Unable to find image 'envoyproxy/envoy-build-ubuntu:09a5a914c904faa39dbc641181cb43b68cabf626' locally
09a5a914c904faa39dbc641181cb43b68cabf626: Pulling from envoyproxy/envoy-build-ubuntu
docker: no matching manifest for linux/arm/v8 in the manifest list entries.
See 'docker run --help'.
```

**Comment by lizan** [comment]
How about `envoyproxy/envoy-build-ubuntu:09a5a914c904faa39dbc641181cb43b68cabf626-arm64`?

**Comment by beriberikix** [comment]
Try as I might, I can't get it build on a Pi 4. I've tried Raspbian and Ubuntu, different versions and I think I hit a OOM issue at some point. Perhaps there's a way to strip out stuff, or free up just a little more RAM?

**Comment by moderation** [comment]
@beriberikix as per https://github.com/envoyproxy/envoy/issues/1861#issuecomment-606021305 did you try restricting to 3 cores with `--jobs 3`. I also outline compiling out all the extensions I don't use.

**Comment by beriberikix** [comment]
@moderation I didn't - I was using the CI scripts since the only prerequisite is Docker. Any more detailed tutorial or walkthrough you recommend?

**Comment by stevesloka** [comment]
I've got a build running now. I'm just using the docker scripts to build. I'm happy to blog on this to share the info that I'm using. I just edited the scripts to pas he `--jobs` arg like @moderation suggested which is super helpful, otherwise, my Pi4 locks up after some time.

**Comment by moderation** [comment]
If you are building on the RPI4, Docker isn't required. I think it makes building Envoy _harder_. See last night's fun thread at https://twitter.com/moderat10n/status/1252445788162060290. I can only attribute the aversion to using Linux and compiling stuff from source to the use of MacOS.

**Comment by stevef1uk** [comment]
> Try as I might, I can't get it build on a Pi 4. I've tried Raspbian and Ubuntu, different versions and I think I hit a OOM issue at some point. Perhaps there's a way to strip out stuff, or free up just a little more RAM?

I have already stopped the k3s.agent and not much else running. I am building on an external hard drive as my SSD card is only 16Mb. 

Last time I tried something this complicated and long winded it was to port Go & docker to Rasbian back in 2014, which I succeeded in doing. I gave up using 'C' in the early 2000s and these days I I have been spoilt in how easy it is to cross compile code written in Go so this is a bit of a 'bad education'.

**Comment by stevesloka** [comment]
Here's how I did my build which works! I may have missed some steps however in the process, so I can update accordingly. Thanks for the tips @moderation! =)

https://stevesloka.com/compile-envoy-on-raspberry-pi4/

**Comment by al45tair** [comment]
@stevesloka Your steps basically work, though I didn't use your YAML to disable components Contour didn't need, and I had to install some additional packages as I was on a clean system. The build then failed on (from memory) a `'\x00' <= c && c <= '\x1f'` expression because char was unsigned, so `'\x00' <= c` is always true and generates a warning. I replaced the entire expression with `static_cast<unsigned char>(c) <= '\x1f'` comparison instead, which has the same effect and made the compiler happy.

**Comment by al45tair** [comment]
OK, so `ci/run_envoy_docker.sh 'ci/do_ci.sh bazel.release.server_only'` was awesome. Much easier. Thanks @lizan.

**Comment by trilom** [comment]
Put this together but it does not build envoy, something about `external/go_sdk/bin` location not being found.  I've been able to nurse it past but have ran into a couple other issues that I believe might be around environment.  Either way this feels pretty close to an alpine build on edge if it's useful to anyone.

```dockerfile
FROM alpine:edge AS bazel

ARG BAZEL_VERSION='3.3.1'

RUN mkdir /bazel
WORKDIR /bazel
RUN wget https://github.com/bazelbuild/bazel/releases/download/${BAZEL_VERSION}/bazel-${BAZEL_VERSION}-dist.zip
RUN unzip bazel-${BAZEL_VERSION}-dist.zip
# script needs bash, build needs rest
RUN apk add bash=5.0.17-r0 \
            openjdk8=8.242.08-r2 \
            build-base=0.5-r2 \
            linux-headers=5.4.5-r1 \
            zip=3.0-r8 \
            python3=3.8.3-r0 
RUN ln -s /usr/bin/python3 /usr/local/bin/python && \
    env JAVA_HOME=/usr/lib/jvm/java-1.8-openjdk EXTRA_BAZEL_ARGS="--host_javabase=@local_jdk//:jdk" bash ./compile.sh

FROM alpine:edge AS envoy
RUN apk add git=2.27.0-r0 \
            openjdk8=8.242.08-r2 \
            autoconf=2.69-r2 \
            automake=1.16.2-r0 \
            libtool=2.4.6-r7 \
            samurai=1.1-r0 \
            cmake=3.17.3-r0 \
            python3=3.8.3-r0 \
            build-base=0.5-r2 \
            llvm10-dev=10.0.0-r2 \
            go=1.14.4-r1 \
            clang=10.0.0-r3 \
            linux-headers=5.4.5-r1 \
            libc6-compat=1.1.24-r9 \
            musl-dev=1.1.24-r9
RUN go get -u github.com/bazelbuild/buildtools/buildifier
RUN go get -u github.com/bazelbuild/buildtools/buildozer
RUN git clone https://github.com/envoyproxy/envoy.git /envoy
WORKDIR /envoy
COPY --from=bazel /bazel/output/bazel /usr/local/bin/bazel

RUN ln -sf /usr/lib/llvm10/ /opt/llvm && \
    ln -sf /usr/bin/python3 /usr/local/bin/python && \
    env JAVA_HOME=/usr/lib/jvm/java-1.8-openjdk \
    bazel build //source/exe:envoy-static --sandbox_debug

```

**Comment by trilom** [comment]
`ci/run_envoy_docker.sh 'ci/do_ci.sh bazel.release.server_only'` no longer works Tick-Tocker/bazel-arm64 does not have Bazel 3.3.1 version and the `.bazelversion` specifies 3.3.1.  If you switch it to 3.3.0 it will build but I am not sure if it is successful yet.

**Comment by lizan** [comment]
@Jingzhao123 ^^

**Comment by lizan** [comment]
@trilom It is not recommended to build a musl based binary, you have to disable signal_actions and gperftools. Disabling latter will impact performance significantly (around 30~40% on x64 on a simple benchmark)

**Comment by trilom** [comment]
> @trilom It is not recommended to build a musl (alpine) based binary, you have to disable signal_actions and gperftools. Disabling latter will impact performance significantly (around 30~40% on x64 on a simple benchmark)

I did read that at one point, I'm not versed in building with musl but I did read similar, and that is for the most part why I stopped with the alpine approach.  Thank you for the information.  My ideal goal is to use something similar to `docker buildx` for this so that I can have some CI to this image ([something similar to this](https://github.com/trailmix/dnsmasqPXE/blob/master/.github/workflows/buildx.yml)), but aside from that the command above should fit my needs.

**Comment by Jingzhao123** [comment]
@lizan @trilom It has been updated.

**Comment by CelsoSantos** [comment]
Realistically speaking, how far are we from an arm64 image that could be considered stable?
From my understanding, we already have an image that runs on arm64, that can be used to build Envoy from arm64 architectures. But then does this produce an arm64 image or just the binary that still needs to be containerized?

Also (though, this one is my fault for not knowing), how long does it take to build an arm64 image with that container? Hours, minutes? Days? I have a RPi4 available that I could use to build Envoy

**Comment by carmiac** [comment]
@CelsoSantos I ran the build just a couple of days ago on my RPi4.  It produced a static non-containerized binary that I was able to drop onto my target device and run.  The build time was around 12 hours, though I did notice that it was just using a single core.

**Comment by moderation** [comment]
@carmiac the single core issue you saw began with a recent release of Bazel. Here is what I posted in the Bazel Slack.

> This line in the Envoy Proxy :bazel: config is the culprit - https://github.com/envoyproxy/envoy/blob/master/.bazelrc#L14. According to the documentation this flag has remained as false as the default. But my compilation went to a single core when I jumped to Bazel 3.2.0. The config in the Envoy code doesn't specify whether it is true or false. My best guess is that in 3.2.0 the interpretation of the flag without a true or false changed and the Envoy Proxy definition flipped to being evaluated as true.  The following override in ~/.bazelrc has me back to using multiple concurrent cores.
```
# disable as this was restricting Bazel actions to 1
build --experimental_local_memory_estimate=false
```
Setting this flag will allow you to use more than one core. I had a 4G RAM RPi4 and I had to restrict the build using `--jobs=3` or `--jobs=2` to prevent memory exhaustion. I just upgraded to am 8G RAM RPi4 and no longer have this issue and can use all 4 cores. All of the above assumes arm64 using a 64 bit OS like Ubuntu Server.

Lastly check out https://stevesloka.com/compile-envoy-on-raspberry-pi4/.

**Comment by al45tair** [comment]
@CelsoSantos It's been a while, but IIRC the command I ran back on the 18th of May worked and generated a Docker image (which I uploaded to Docker Hub as al45tair/envoy-arm64). It seems to work.

@moderation @carmiac It's probably worth getting some time on an ARM64 server machine rather than building on a Pi. Even if you don't have access to one at work (I'm lucky in that regard), you can spin them up in AWS — there are Graviton and Graviton2 instances available, I believe — and they'll build it a lot faster than a Pi will.

**Comment by moderation** [comment]
@al45tair I'm pretty happy with my local RPi4 dev environment. After the initial build the subsequent incremental builds are relatively quick and compiling out all the extensions you don't need works well. I'm in a very small minority of people who don't think Docker helps with anything so I'm building Bazel and Envoy straight on the RPi4

**Comment by lizan** [comment]
> Realistically speaking, how far are we from an arm64 image that could be considered stable?

I wouldn't be comfortable saying it is stable until it passes most of our test suite on arm64.

**Comment by vielmetti** [comment]
A note that Bazel 3.4.0 now ships with official arm64 binaries - 

https://github.com/bazelbuild/bazel/issues/8833#issuecomment-657698507

That may save you a build step, @moderation , if you can adapt to any changes associated with the Bazel point release at 3.4.0.

**Comment by moderation** [comment]
Thanks @vielmetti. Looks like they need to work on their automation a bit as the latest release 3.4.1 isn't yet available. But definitely looks promising
